woocommerce/google-listings-and-ads
# Shipping Integration - Pre-fill shipping rates

### Changes proposed in this Pull Request:



Implements part of #1117.

This PR adds a new API (`mc/shipping/rates/suggestions`) that returns a list of suggested shipping rates for the given country. There is also another "batch" API under the same namespace (`mc/shipping/rates/suggestions/batch`) that accepts multiple country codes as input and returns suggested shipping rates for each of them in a batch response.

The API parses the WooCommerce shipping zones to fetch the list of shipping methods for each country and returns a rate only if there is a "Flat-rate" shipping method with a simple cost associated with the country.

The flat rate shipping method can accept a mathematical expression or formula as its "cost" parameter, which can include various variables such as item quantity and order total cost. Since Google does not support adding such formulas as shipping cost, we simply ignore those methods and only return the ones that have a simple numeric value defined as their cost. The same also applies to the "pickup" shipping method.

Even though we parse all of the supported code WooCommece shipping methods in the `ShippingZone` service (`free_shipping`, `local_pickup`, and `flat_rate`) and return them all in the `get_shipping_methods_for_country` method, the API only supports returning a `flat_rate` for a country. This means that we do not return any other shipping methods even if they exist. This is because the current shipping rate UI does not allow setting the "local pickup" or "free shipping" method separately for countries. Once we have the UI design ready, we can change the API to return the data in the desired format for the UI.

If a country has multiple instances of the same shipping method, we always return the one that has a higher cost. Cost is defined differently in the shipping methods. In the "flat rate" method, "cost" is the flat shipping fee that is defined by the user. For the "free shipping" method, however,  we always return the one that has a higher "minimum order amount" defined.

With this change, if a country has no shipping methods defined, it will no longer be returned as an option for the `mc/target_audience/suggestions` API.

Moreover, if the cost is null or undefined for a flat-rate or local-pickup shipping method, the method will not be returned as an option.

These APIs are used to pre-populate the shipping rate inputs presented to the user during onboarding.

### Detailed test instructions:

1. Define multiple shipping zones each containing various regions and shipping methods. Make sure to have at least one country that is defined in multiple zones and has a similar shipping method in each with a different cost.
2. Make a `GET` request to the `mc/shipping/rates/suggestions` API with a `country_code` parameter and confirm the rate that is returned is the same that you defined in your shipping zone settings.
3. Make a `GET` request to the `mc/shipping/rates/suggestions` API and send over the country code that is duplicated in multiple zones as the `country_code` parameter and check the rate is set to the higher one.
4. Make a `GET` request to the `mc/shipping/rates/suggestions/batch` API with multiple country codes defined as the `country_codes` parameter and confirm that the correct rate is returned for all of them.
5. Run the unit tests and confirm they pass.


### Changelog entry

> Add - Pre-fill shipping rates during free listing configuration wizard
